src/vs/workbench/api/common/extHostNotebook.ts

```diff
@@ -372,6 +375,134 @@ export class ExtHostNotebookController implements ExtHostNotebookShape {
 		return fileStats;
 	}
 
+	/**
+	 * Search for query in all notebooks that can be deserialized by the serializer fetched by `handle`.
+	 *
+	 * @param handle used to get notebook serializer
+	 * @param filenamePattern the filename pattern to match files that can be deserialized
+	 * @param textQuery the text query to search using
+	 * @param token cancellation token
+	 * @returns `IRawClosedNotebookFileMatch` for every file. Files without matches will just have a `IRawClosedNotebookFileMatch`
+	 * 	with no `cellResults`. This allows the caller to know what was searched in already, even if it did not yield results.
+	 */
+	async $searchInNotebooks(handle: number, filenamePattern: string[], textQuery: ITextQuery, token: CancellationToken): Promise<{ results: IRawClosedNotebookFileMatch[]; limitHit: boolean }> {
+		const serializer = this._notebookSerializer.get(handle)?.serializer;
+		if (!serializer) {
+			return {
+				limitHit: false,
+				results: []
+			};
+		}
+
+		const runFileQueries = async (includes: (string)[], token: CancellationToken, textQuery: ITextQuery): Promise<URI[]> => {
+			const uris = new ResourceSet();
+			await Promise.all(includes.map(include => {
+				const query: IFileQuery = {
+					type: QueryType.File,
+					filePattern: include,
+					folderQueries: textQuery.folderQueries,
+					maxResults: textQuery.maxResults,
+				};
+				return this._extHostSearch.$doInternalFileSearchWithCustomCallback(query, token, (data) => {
+					data.forEach(e => {
+						uris.add(URI.revive(e));
+					});
+				});
+			}));
+
+			return Array.from(uris.keys());
+		};
+
+		const filesToScan = await runFileQueries(filenamePattern, token, textQuery);
+
+		const results = new ResourceMap<INotebookFileMatchNoModel>();
+		let limitHit = false;
+		const promises = filesToScan.map(async (uri) => {
+			const cellMatches: INotebookCellMatchNoModel[] = [];
+
+			try {
+				if (token.isCancellationRequested) {
+					return;
+				}
+				if (textQuery.maxResults && [...results.values()].reduce((acc, value) => acc + value.cellResults.length, 0) > textQuery.maxResults) {
+					limitHit = true;
+					return;
+				}
+
+				const simpleCells: Array<{ input: string; outputs: string[] }> = [];
+				const notebook = this._documents.get(uri);
+				if (notebook) {
+					const cells = notebook.apiNotebook.getCells();
+					cells.forEach(e => simpleCells.push(
+						{
+							input: e.document.getText(),
+							outputs: e.outputs.flatMap(value => value.items.map(output => output.data.toString()))
+						}
+					));
+				} else {
+					const fileContent = await this._extHostFileSystem.value.readFile(uri);
+					const bytes = VSBuffer.fromString(fileContent.toString());
+					const notebook = await serializer.deserializeNotebook(bytes.buffer, token);
+					const data = typeConverters.NotebookData.from(notebook);
```

Comment:
check `token.isCancellationRequested` right after the `await` so we don't have to do the heavy lifting work below if it's already cancelled.